subprojects/model-core/src/integTest/groovy/org/gradle/api/provider/ProviderConventionMappingIntegrationTest.groovy

```diff
@@ -51,26 +51,26 @@ class ProviderConventionMappingIntegrationTest extends AbstractIntegrationSpec {
         expectDocumentedFailure()
     }
 
-    def "emits deprecation warning when convention mapping is used with Property"() {
+    def "convention mapping can be used with Property"() {
         buildFile << """
             abstract class MyTask extends DefaultTask {
                 @Internal abstract Property<String> getFoo()
+                @Internal abstract Property<String> getBar()
 
                 @TaskAction
                 void useIt() {
-                    // convention mapping for Property is already ignored
-                    assert foo.get() == "other"
+                    assert foo.get() == "foobar"
+                    assert bar.get() == "foobar"
                 }
             }
             tasks.register("mytask", MyTask) {
-                conventionMapping.map("foo", { project.objects.property(String).convention("foobar") })
-                foo.convention("other")
+                conventionMapping.map("foo") { "foobar" }
+                conventionMapping.map("bar") { providers.provider { "foobar" } }
```

Comment:
This is close to what I was thinking, but we don't need to support conventions that return a Provider.

We also don't want to support _new_ cases of using convention mapping with existing `Property` properties.

So if I have:
```
abstract class MyTask extends DefaultTask {
    @Internal String getFoo() { ... }
    void setFoo(String foo) { ... }
    @Internal abstract Property<String> getBar()
}
```

I should be able to do:
```
conventionMapping.map("foo", { "foobar" })
```
I should _not_ be able to do:
```
conventionMapping.map("bar", { "foobar" })
```

However, when my task has been upgraded to use a `Property<String>` for `foo`, then:
```
conventionMapping.map("foo", { "foobar" })
```
should continue to work and mean the same as:
```
foo.convention(providers.provider { "foobar" })
```

Reply:
OK, so two things:

1) Remove support for convention values that are themselves `Provider`s.

I don't mind doing this, should be relatively easy.

2) Only support wiring `ConventionMapping.map()` through to `Property.convention()` for upgraded code.

We were discussing this, I see three ways:

a) Ignore the problem altogether, and support wiring through for all `Property` instances. All such usage is "illegal" as it's using internal APIs anyway, and we'll soon deprecate doing so in Gradle 9.0.

b) Do it properly, and observe if it's old code or new code that is calling `ConventionMapping.map()` somehow (reflection + bytecode magicks).

c) Allow anyone to call `ConventionMapping.map()` for properties tagged with `@UpgradedProperty`.

I think I'd go for c) if not a). Investing in b) seems not worth the trouble. The intention is clear: use `Property.convention()`, not `ConventionMapping`, wherever you already can. If you accidentally miss that advice, you'll have to deal with the deprecation of `ConventionMapping`, and do it with `Property.convention()` after that. It doesn't sound like a problem many folks are likely to fall into.

WDYT?

Reply:
I don't have a strong feeling about it.  Maybe we could make it so upgraded code gets a specialized ConventionMapping that allows convention mapping with `Property`, but the normal one does not?

Reply:
I discussed this with @adammurdoch. We came up with the solution to allow this to work for any lazy `Property`. If we see a problem with plugin developers misusing it, we can always add some further checks. However, we expect no such trouble, and thus would like to save time on not implementing these further checks up front.

I've also removed the ability to set a convention for `Property` with a `Provider` value.